feat: allow customizing setting tokens through the interceptor

duizendnegen opened the pull request:
Allow other ways of setting tokens. In our application, the token is sent in the form `{ 'x-auth-token': token }`, not the default bearer token.

This might catch a few users by surprise, as so far it's been in the docs to _implement_ (rather than _extend_) the AuthService. I'd recommend creating bumping a minor version, not releasing this as a patch.

serhiisol commented:
Hi @duizendnegen, thanks for the PR, that's definitely good addition to the library. But what if we'll adjust it a little bit and will add small snippet to the auth service itself, which will provide default header by default ? E.g.

```typescript
export abstract class AuthService {
  // ...

  public abstract addToken(req: HttpRequest<any>): Observable<HttpRequest<any>>;

  // ...
}

@Injectable()
export class AuthInterceptor implements HttpInterceptor {
  // ...

  private addToken(req: HttpRequest<any>): Observable<HttpRequest<any>> {
    const authService: AuthService =
      this.injector.get<AuthService>(AUTH_SERVICE);

    if (typeof authService.addToken === 'function') {
      return authService.addToken(req);
    }

    return first(map(
      authService.getAccessToken(),
      (token: string) => {
        if (token) {
          return req.clone({
            setHeaders: { Authorization: `Bearer ${token}` }
          });
        }

        return req;
      }
    ));
  }

  // ...
}
```

In this case no-one would need to adjust anything in the current setup. Less breaking changes :)

duizendnegen commented:
Excellent suggestion - I agree. Implemented that now, however, nested inside the `first(map( ... ))`, that brings down the responsibilities of the `AuthenticationService` itself. Added docs to the README for how to use.
Let me know if I should squash this still.

serhiisol commented:
Nice, but could you fix `extends` back to `implements`, plus adjust the things accordingly ?

duizendnegen commented:
Yep, done.
I thought to keep those there anyway as extending would be more flexible than implementing for future additions, but any future addition would be breaking as per semver anyway, so :+1:

serhiisol commented on src/auth.service.ts:
```diff
@@ -69,4 +69,19 @@ export abstract class AuthService {
+   *
+   * @returns {[name: string]: string | string[]}
+   */
+  public getHeaders?(token: string) : { [name: string]: string | string[] };
```
could you add `abstract` here ?

duizendnegen replied:
Done

serhiisol commented on README.md:
````diff
@@ -115,3 +115,7 @@ by ```ProtectedGuard``` and won't be authenticated
+
+### Customizing authentication headers
+
+By default, requests are intercepted and a ```{ Authorization: 'Bearer ${token}'}``` header is injected. To customize this behavior, implement the ```getHeaders``` method on your ```AuthenticationService```
````
could you method interface to the class interface above?

duizendnegen replied:
Decided not to do so: don't want to give the idea that this is mandatory to implement, I think it's really "advanced configuration"

serhiisol replied:
ok, sounds reasonable

duizendnegen commented:
Done, let me know if there's anything else.

serhiisol commented:
Ok, nice, merged. But I forgot to ask you to bump the version, but that's not a problem, going to do that tonight. Hold on 👍
